storage/memetcd/peer_store.go

```diff
@@ -0,0 +1,535 @@
+		})
+	}
+
+	validcfg.MemoryStore = cfg.MemoryStore.Validate()
+
+	return validcfg
+}
+
+// New creates a new PeerStore backed by memory.
+func New(provided Config) (storage.PeerStore, error) {
+	cfg := provided.Validate()
+
+	etcdClient, err := clientv3.New(cfg.EtcdConfig)
+	if err != nil {
+		return nil, err
+	}
+	etcdClient.KV = namespace.NewKV(etcdClient.KV, cfg.EtcdNamespace)
+	etcdClient.Watcher = namespace.NewWatcher(etcdClient.Watcher, cfg.EtcdNamespace)
+	etcdClient.Lease = namespace.NewLease(etcdClient.Lease, cfg.EtcdNamespace)
+
+	memoryStore, err := memory.New(cfg.MemoryStore)
+	if err != nil {
+		return nil, err
+	}
+
+	ps := &peerStore{
+		cfg:         cfg,
+		etcd:        etcdClient,
+		memoryStore: memoryStore,
+		closed:      make(chan struct{}),
+	}
+
+	if err = ps.Bootstrap(); err != nil {
+		return nil, err
+	}
+
+	return ps, nil
+}
+
+type serializedPeer []byte
+
+func newPeerKey(p *bittorrent.Peer) serializedPeer {
+	b := make([]byte, 20+2+len(p.IP.IP))
+	copy(b[:20], p.ID[:])
+	binary.BigEndian.PutUint16(b[20:22], p.Port)
+	copy(b[22:], p.IP.IP)
+	return b
+}
+
+const (
+	EtcdKeySep 					= "/"
+	EtcdAnnouncePrefix 			= "announce"
+	EtcdAnnounceSeederPrefix 	= "s" 				//len 1
+	EtcdAnnounceLeecherPrefix	= "l"				//len 1
+)
+
+func newAnnounceKey(ih *bittorrent.InfoHash, p *bittorrent.Peer, seeder bool) ([]byte) {
+	prefix, sep := []byte(EtcdAnnouncePrefix), []byte(EtcdKeySep)
+	var peerType []byte
+	if seeder {
+		peerType = []byte(EtcdAnnounceSeederPrefix)
+	} else {
+		peerType = []byte(EtcdAnnounceLeecherPrefix)
+	}
+
+	var pk serializedPeer
+	if p != nil {
+		pk = newPeerKey(p)
+	}
+
+	key := make([]byte, 0, len(prefix) + len(sep) * 3 + len(ih) + len(peerType) + len(pk))
+	key = append(key, prefix...)
+	key = append(key, sep...)
+	key = append(key, ih[:]...)
+	key = append(key, sep...)
+	key = append(key, peerType...)
+	key = append(key, sep...)
+	key = append(key, pk...)
+
+	return key
+}
+
+func decodeAnnounceKey(key []byte) (*bittorrent.InfoHash, []byte, bool) {
+	prefix, sep := []byte(EtcdAnnouncePrefix), []byte(EtcdKeySep)
+
+	offset := len(prefix) + len(sep)
+	ih := bittorrent.InfoHashFromBytes(key[offset:offset+20])
+
+	offset += len(ih) + len(sep)
+	peerType := key[offset:offset + 1]
+	var seeder bool
+	if string(peerType) == EtcdAnnounceSeederPrefix {
+		seeder = true
+	} else {
+		seeder = false
+	}
+
+	offset += 1 + len(sep)
+	peerKey := key[offset:]
+
+	return &ih, peerKey, seeder
+}
+
+type peerStore struct {
+	cfg         Config
+	etcd        *clientv3.Client
+	memoryStore storage.PeerStore
+	closed      chan struct{}
+}
+
+var _ storage.PeerStore = &peerStore{}
+
+func (ps *peerStore) PutSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	if err := ps.memoryStore.PutSeeder(ih, p); err != nil {
+		return err
+	}
+
+	if ps.cfg.EtcdOpAsync {
+		go func() {
```

Comment:
Was starting out in golang and not comfortable with channels when I wrote this feature so rationalized everything around it :P
Granted there will be races here but bittorrent clients will not get affected since tracker serves responses of announce and scrape from in-memory. in case of process restart, swarm will not be consistent. more like a best-effort which again should be alright mostly.

Will think more on this

Reply:
So race only matters if it is for same (infohash,peer) which honestly should be a rare event unless etcd is down and etcd get/put/delete start blocking for a long time or torrent clients are misbehaving.
To solve for race means channels which if bounded are going to cause backpressure. Definitely not want that because we are favoring performance (http request should return asap) over etcd conistency. The other option is to use an unbounded queue. As you suggested or shard channels/queues on basis of ih,k. All of these introduce complexity which does not seem worth in on first glance. Still I will make a note of this in comments and leave this as it is for now

Reply:
sure, make sense! But we should see(with some experiment) if it might hamper bittorrent p2p connections(may be due to longer timeout, not sure though) if a peer from announce list not exist.